homeassistant/components/volvooncall/__init__.py

```diff
@@ -215,42 +188,87 @@ def instrument(self, vin, component, attr, slug_attr):
 
     def vehicle_name(self, vehicle):
         """Provide a friendly name for a vehicle."""
-        if (
-            vehicle.registration_number and vehicle.registration_number.lower()
-        ) in self.names:
-            return self.names[vehicle.registration_number.lower()]
-        if vehicle.vin and vehicle.vin.lower() in self.names:
-            return self.names[vehicle.vin.lower()]
-        if vehicle.registration_number:
+        if vehicle.registration_number and vehicle.registration_number != "UNKNOWN":
             return vehicle.registration_number
         if vehicle.vin:
             return vehicle.vin
-        return ""
+        return "Volvo"
+
+    def discover_vehicle(self, vehicle):
+        """Load relevant platforms."""
+        self.vehicles.add(vehicle.vin)
+
+        dashboard = vehicle.dashboard(
+            mutable=self.config[CONF_MUTABLE],
+            scandinavian_miles=self.config[CONF_SCANDINAVIAN_MILES],
+        )
+
+        for instrument in (
+            instrument
+            for instrument in dashboard.instruments
+            if instrument.component in PLATFORMS
+        ):
 
+            self.instruments.add(instrument)
 
-class VolvoEntity(Entity):
+            self.hass.async_create_task(
+                discovery.async_load_platform(
+                    self.hass,
+                    PLATFORMS[instrument.component],
+                    DOMAIN,
+                    (
+                        vehicle.vin,
+                        instrument.component,
+                        instrument.attr,
+                        instrument.slug_attr,
+                        self.coordinator,
+                    ),
+                    self.config,
+                )
+            )
+
+    async def update(self):
+        """Update status from the online service."""
+        if not await self.connection.update(journal=True):
+            _LOGGER.warning("Could not query server")
+            return False
+
+        for vehicle in self.connection.vehicles:
+            if vehicle.vin not in self.vehicles:
+                self.discover_vehicle(vehicle)
+
+        # this is currently still needed for device_tracker, which isn't using the update coordinator yet
+        async_dispatcher_send(self.hass, SIGNAL_STATE_UPDATED)
+
+        return True
+
+
+class VolvoEntity(CoordinatorEntity):
     """Base class for all VOC entities."""
 
-    def __init__(self, data, vin, component, attribute, slug_attr):
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        vin,
+        component,
+        attribute,
+        slug_attr,
+        coordinator,
+    ):
         """Initialize the entity."""
-        self.data = data
+        super().__init__(coordinator)
+
+        self.hass = hass
```

Comment:
Don't pass in `hass`. It will be set on the entity when the entity has been added to home assistant. This means you can access `hass` with `self.hass`, after the entity has been added to home assistant.